### PR TITLE
feat: support provisioned concurrency and proactive initialization

### PR DESCRIPTION
**Pull request by astuyve**
### Motivation

This PR fixes cold start tags for provisioned concurrency and proactive initialization. It also fixes the cold start span location:
image

Note that the span is correctly placed beginning where the [sandbox initialized](https://ddserverless.datadoghq.com/apm/trace/5812341648069855126?colorBy=service&env=staging&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=5812341648069855126&spanViewType=metadata&timeHint=1683642323273.666)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/serverless/executioncontext/executioncontext.go [hunk 1]
```diff
@@ -72,13 +78,34 @@ func (ec *ExecutionContext) SetFromInvocation(arn string, requestID string) {
 	ec.arn = strings.ToLower(arn)
 	ec.lastRequestID = requestID
 	if len(ec.coldstartRequestID) == 0 {
-		ec.coldstart = true
+		// Putting this within the guard clause of
+		// coldstartRequestID == 0 ensures this is set once
+		// 10 seconds is the maximimum init time, so if the difference
+		// between the first invocation start time and initTime
+		// is greater than 10, we're in a proactive_initialization
+		// not a cold start
```
Comment (purple4reina) [inline pkg/serverless/executioncontext/executioncontext.go:86]: Question: is there a chance that a proactive init can happen within less than 10s if the initialization is very very fast? In that case, what does the trace look like?

File: pkg/serverless/tags/tags.go [hunk 2]
```diff
@@ -173,8 +173,13 @@ func buildTags(tags map[string]string, tagsToSkip []string) map[string]string {
 }
 
 // AddColdStartTag appends the cold_start tag to existing tags
-func AddColdStartTag(tags []string, coldStart bool) []string {
-	tags = append(tags, fmt.Sprintf("cold_start:%v", coldStart))
+func AddColdStartTag(tags []string, coldStart bool, proactiveInit bool) []string {
+	if proactiveInit {
+		tags = append(tags, fmt.Sprintf("cold_start:%v", false))
```
Comment (purple4reina) [inline pkg/serverless/tags/tags.go:178]: why not just?

```suggestion
		tags = append(tags, "cold_start:false")
```

Comment (purple4reina) [reply pkg/serverless/tags/tags.go:178]: And similarly with the two below as well.

File: pkg/serverless/logs/parse.go [hunk 3]
```diff
@@ -24,13 +24,19 @@ type platformObjectRecord struct {
...
+// early in the invocation, which is a bit confusing
+// TODO Astuyve - refactor out initDurationMs to draw from TelemetryAPI
 type reportLogMetrics struct {
 	durationMs            float64
 	billedDurationMs      int
 	memorySizeMB          int
 	maxMemoryUsedMB       int
 	initDurationMs        float64
 	initDurationTelemetry float64
+	initStartTime         time.Time
```
Comment (purple4reina) [inline pkg/serverless/logs/parse.go:39]: nit: _technically_ I feel like several of these fields (including the one you added) should be moved to their own separate struct. It seems to me that the `startLogItem`, `runtimeDoneItem`, and `reportLogItem` were initially intended to contain information from a single type of log event.

This doesn't need to be fixed. But it did confuse me for a second.

**Review by purple4reina — APPROVED** [review]
I will admit that I do not 100% grok this code, but I'm not seeing any issues with it, no matter how deeply I look. I feel like there should be ways to make the code logic easier to understand during a first time read, but I don't know what they are and I'm not sure it's worth our time anyway.

File: pkg/serverless/logs/parse.go (hunk 3 above)
Comment (astuyve) [reply pkg/serverless/logs/parse.go:39]: You're 10000% right here. We get the initDuration in MS via the report log, which is what this struct was meant to do. I extended it unnaturally when adding support for the telemetry API field here, but I do think I should move it.

**Pull request merged**